// @ngx-pwa/local-storage/esm2015/lib/validation/json-schema.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Avoid: will be removed in v8 in favor of another interface.
 * @record
 */
export function JSONSchemaConst() { }
if (false) {
    /**
     * Checks if a value is strictly equal to this.
     * Can't be an object or array, as two objects or arrays are never equal.
     * @type {?}
     */
    JSONSchemaConst.prototype.const;
}
/**
 * Avoid: will be removed in v8 in favor of another interface.
 * @record
 */
export function JSONSchemaEnum() { }
if (false) {
    /**
     * Checks if a value is strictly equal to one of the value of enum.
     * Can't be an object or array, as two objects or arrays are never equal.
     * @type {?}
     */
    JSONSchemaEnum.prototype.enum;
}
/**
 * @record
 */
export function JSONSchemaBoolean() { }
if (false) {
    /**
     * Type for a boolean value.
     * @type {?}
     */
    JSONSchemaBoolean.prototype.type;
}
/**
 * Avoid: will be removed in v8.
 * @record
 */
export function JSONSchemaNull() { }
if (false) {
    /**
     * Type for a null value.
     * @type {?}
     */
    JSONSchemaNull.prototype.type;
}
/**
 * @record
 */
export function JSONSchemaString() { }
if (false) {
    /**
     * Type for a string value.
     * @type {?}
     */
    JSONSchemaString.prototype.type;
    /**
     * Maxium length for a string.
     * Must be a non-negative integer.
     * @type {?|undefined}
     */
    JSONSchemaString.prototype.maxLength;
    /**
     * Minimum length for a string.
     * Must be a non-negative integer.
     * @type {?|undefined}
     */
    JSONSchemaString.prototype.minLength;
    /**
     * Pattern to match for a string.
     * Must be a valid regular expression, WITHOUT the / delimiters.
     * @type {?|undefined}
     */
    JSONSchemaString.prototype.pattern;
}
/**
 * @record
 */
export function JSONSchemaNumeric() { }
if (false) {
    /** @type {?} */
    JSONSchemaNumeric.prototype.type;
    /**
     * Check if a number is a multiple of x.
     * Must be strictly greater than 0.
     * @type {?|undefined}
     */
    JSONSchemaNumeric.prototype.multipleOf;
    /**
     * Check if a number is less or equal than this maximum.
     * @type {?|undefined}
     */
    JSONSchemaNumeric.prototype.maximum;
    /**
     * Check if a number is strictly less than this maximum.
     * @type {?|undefined}
     */
    JSONSchemaNumeric.prototype.exclusiveMaximum;
    /**
     * Check if a number is greater or equal than this minimum.
     * @type {?|undefined}
     */
    JSONSchemaNumeric.prototype.minimum;
    /**
     * Check if a number is strictly greater than this minimum.
     * @type {?|undefined}
     */
    JSONSchemaNumeric.prototype.exclusiveMinimum;
}
/**
 * @record
 */
export function JSONSchemaArray() { }
if (false) {
    /**
     * Type for an array value.
     * Will be *required* in v8, so explicit this now!
     * @type {?|undefined}
     */
    JSONSchemaArray.prototype.type;
    /**
     * Schema for the values of an array.
     * Avoid to use an array of schemas, this feature will be removed in v8.
     * @type {?}
     */
    JSONSchemaArray.prototype.items;
    /**
     * Check if an array length is less or equal to this value.
     * Must be a non negative integer.
     * @type {?|undefined}
     */
    JSONSchemaArray.prototype.maxItems;
    /**
     * Check if an array length is greater or equal to this value.
     * Must be a non negative integer.
     * @type {?|undefined}
     */
    JSONSchemaArray.prototype.minItems;
    /**
     * Check if an array only have unique values.
     * @type {?|undefined}
     */
    JSONSchemaArray.prototype.uniqueItems;
}
/**
 * @record
 */
export function JSONSchemaObject() { }
if (false) {
    /**
     * Type for an object value.
     * Will be *required* in v8, so explicit this now!
     * @type {?|undefined}
     */
    JSONSchemaObject.prototype.type;
    /**
     * List of properties schemas for an object.
     * @type {?}
     */
    JSONSchemaObject.prototype.properties;
    /**
     * Array of names of the required properties for an object.
     * Properties set as required should be present in 'properties' too.
     * Note that in the last spec, booleans are not supported anymore.
     * @type {?|undefined}
     */
    JSONSchemaObject.prototype.required;
}
